docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+    itself can configured (and fine-tuned) elsewhere (also the app?, operators?)
+* Two options for introducing several TX lists:
+  1. Multiple `Clist`s inside `CListMempool`, the cache stays inside `CListMempool`.
+  2. Multiple `CListMempool` instances, the cache needs to be outside `CListMempool` (likely, directly controlled by the reactor)
+* We need to investigate the details of these two options in terms of complexity and risks in the code
+
+ADR: Probably part of [Data flow](#data-flow) section:
+1. some bullets are clearly "Entry flow"
+2. other are general considerations of all flows, or architectural.
+   These are likely to go in a first, introductory subsection of [Data flow](#data-flow) section
+
+### Data flow
+
+* Entry flow:
+  * [Rough description]: the steps when a Tx enters the node are unmodified, except
+    * `CheckTx` (optionally) returns lane information (if TX valid)
+    * The lane information is used to decide to which transaction list the Tx will be added
+* Broadcast flow:
+  * Higher-priority lanes get to be gossiped first, and gossip within a lane is still in FIFO order.
+  * [Rough description]: we need to extend the loop in the broadcast goroutine. Several channels now, not just one
+    * To decide
+      * if we're going for (simpler) solution where with max 2 lanes,
+        the broadcast goroutine itself could manage the priority of lanes within the existing `select` at the end of the loop
+      * if we're going for (future-proof) solution where max lanes can be configurable,
+        we need to refactor method `TxWaitChan` (or the code that sends the data to that channel) to manage the lane priority there.
+        In this case, the broadcast goroutine will be mainly unmodified.
+* Reap flow
+  * TXs are reaped from higher-priority lanes first, in FIFO order.
+  * We go from the reap loop (currently just one), to probably a nested one
+    * the outer loop goes through all TX lists (remember, one per lane), in decreasing priority order
+    * the current `break` statements in the inner loop (limit of bytes or gas reached), should also break from the outer loop
+* Exit flow (Update, Re-Check). Update: unconditionally; Re-Check: only if App says TX is now invalid
+  * Transactions in higher-priority lanes are processed (updated) first.
+  * Depends on how the multiple TX lists are implemented (see discussion above)
+  * Update contains Re-check at the end
+  * Remember Update is done while holding the mempool lock (no new CheckTx calls... they have to wait)
+  * Changes:
+    * We could update the different TX lists in parallel
+    * `Update` method (implementing `Update` method in `Mempool` interface) is implemented by `CListMempool` so
+      * If we go for several `CListMempool`s, then we'll have to call `Update` on each of them with all the TXs every time
+        * or move `Update` method out of `CListMempool` (where?)
```

Comment:
We probably need an actor to manage the multiple lanes (or mempools).

Reply:
Maybe... we didn't get to that level of detail yet